src/client-side-encryption/providers/azure.ts

```diff
@@ -0,0 +1,171 @@
+import { type Document } from 'bson';
+
+import { MongoCryptAzureKMSRequestError, MongoCryptKMSRequestNetworkTimeoutError } from '../errors';
+import { type KMSProviders } from '.';
+import * as utils from './utils';
+
+const MINIMUM_TOKEN_REFRESH_IN_MILLISECONDS = 6000;
+
+/**
+ * The access token that libmongocrypt expects for Azure kms.
+ */
+interface AccessToken {
+  accessToken: string;
+}
+
+/**
+ * The response from the azure idms endpoint, including the `expiresOnTimestamp`.
+ * `expiresOnTimestamp` is needed for caching.
+ */
+interface AzureTokenCacheEntry extends AccessToken {
+  accessToken: string;
+  expiresOnTimestamp: number;
+}
+
+/**
+ * @internal
+ */
+export class AzureCredentialCache {
```

Comment:
From slack: Let's capture the deduplication of this in relation to future OIDC efforts